_posts/2025-06-24-kafka-queue.md

```diff
@@ -0,0 +1,416 @@
+
+![KafkaShareConsumer API](/assets/images/posts/2025-06-24-kafka-queue-08.png)
+
+<i>Acknowledging individual records</i>
+
+This allows users to acknowledge an individual record depending on the outcome of the processing of the record. Each record is acknowledged using a call to <code>acknowledge()</code> which takes different types of acknowledgement as an argument: ACCEPT, RELEASE and REJECT.
+This aligns with the actions that can be taken by share group consumers mentioned previously.
+
+In the next example, I have used the new consumer configuration added for this feature, <b>share.acknowledgement.mode</b> and set it to "explicit". This configuration is set to "implicit" by default, which is why I didn’t need to set this configuration for the previous example. 
+
+![KafkaShareConsumer API](/assets/images/posts/2025-06-24-kafka-queue-09.png)
+
+In this example, the consumer will release the record if its key is set to "skip", just to demonstrate the retry of delivery attempts. This record will be fetched again in the next call to <code>poll()</code> and eventually will hit the maximum delivery attempts and transition into <b>Archived</b> state because the client never processes it. Once archived, this record will no longer be available for another delivery. 
+
+Also when a record value cannot be mapped to a valid JSON object, the processing method throws an <code>JsonProcessingException</code> and the client handles it as a poison message. The Kafka Queue feature currently doesn’t support dead letter queue but it will potentially be added in the future. Right now, users have to implement this themselves, similar to this example. Once a poison message record is sent to the dead letter queue topic, the consumer client will reject it so that it gets archived immediately. 
+
+If the client encountered an error not caused by the JSON mapping, it releases the record for another attempt because it could be a transient failure.
+
+When no exception occurred during the processing, the consumer client accepts the record. Once all the records in the batch are processed and acknowledged individually, these states are stored locally in the consumer. Then the client has to call <code>commitSync()</code> or <code>commitAsync()</code> to commit the state to the internal topic, <code>__shared_group_state</code>. If any of the records in the batch hits an error or skips the processing, the client commits the state at that point. The records in the batch that were not processed or acknowledged yet, will be presented to the consumer client again as part of the same acquisition, therefore their delivery count will not be incremented.
+
+<i>Inspecting __share_group_state topic</i>
```

Comment:
Should this be a heading? Looking at the [preview](https://deploy-preview-509--strimzi.netlify.app/blog/2025/06/24/kafka-queue/) it's getting a little lost